File: src/lessons/sort/baseball/InsertBaseballEntity.c
//RemoteBaseball

void out(char* msg);
int getHoleInsert();
void moveInsert(int pos);
int getPlayerColorInsert(int pos);

/* BEGIN TEMPLATE */
void run() {
	/* BEGIN SOLUTION */
	/* Bring the hole in 0,1 */
	if (getHoleInsert() == 0) // It is already on base 0, but on another position
		moveInsert(1);
	while (getHoleInsert() > 1)
		moveInsert(getHoleInsert()-1);
	int player;
	for (player = 2; player < getBasesAmount()*getPositionsAmount(); player ++) {
		//out("Sort player "+player);

		//out("Compare "+(getHole()+1)+":"+getPlayerColor(getHole()+1)+" < "+(getHole()-1)+":"+getPlayerColor(getHole()-1));
		while (getHoleInsert()>0 && getPlayerColorInsert(getHoleInsert()+1) < getPlayerColorInsert(getHoleInsert()-1)) {
			int center = getHoleInsert();// ...2x1... with ascending positions from left to right
			moveInsert(center+1);        // ...21x...
			moveInsert(center-1);        // ...x12...
		}
		while (getHoleInsert() != player)
			moveInsert(getHoleInsert()+1);
	}
	assertSorted("insertion sort");
	/* END SOLUTION */
}
/* END TEMPLATE */

/* BEGIN HIDDEN */

int getPlayerColorInsert(int pos) {
	return getPlayerColor(pos / getPositionsAmount(), pos % getPositionsAmount());
}

void moveInsert(int pos) {
	move(pos / getPositionsAmount(), pos % getPositionsAmount());
}

int getHoleInsert() {
	return getPositionsAmount()*getHoleBase()+getHolePosition();
}

void out(char* msg) {
	if (isSelected()) {
		printf("100 %s\n",msg);
		fflush(stdout);
	}

}
/* END HIDDEN */

File: src/lessons/sort/baseball/InsertBaseballEntity.java
package lessons.sort.baseball;

import lessons.sort.baseball.universe.BaseballEntity;

public class InsertBaseballEntity extends BaseballEntity {

  /* BEGIN TEMPLATE */
  public void run()
  {
    /* BEGIN SOLUTION */
    /* Bring the hole in 0,1 */
    if (_getHole() == 0) // It is already on base 0, but on another position
      _move(1);
    while (_getHole() > 1)
      _move(_getHole() - 1);

    for (int player = 2; player < getBasesAmount() * getPositionsAmount(); player++) {
      // out("Sort player "+player);

      // out("Compare "+(getHole()+1)+":"+getPlayerColor(getHole()+1)+" < "+(getHole()-1)+":"+getPlayerColor(getHole()-1));
      while (_getHole() > 0 && _getPlayerColor(_getHole() + 1) < _getPlayerColor(_getHole() - 1)) {
        int center = _getHole(); // ...2x1... with ascending positions from left to right
        _move(center + 1);       // ...21x...
        _move(center - 1);       // ...x12...
      }
      while (_getHole() != player)
        _move(_getHole() + 1);
    }
    assertSorted("insertion sort");
  }

  int _getPlayerColor(int pos) { return getPlayerColor(pos / getPositionsAmount(), pos % getPositionsAmount()); }
  void _move(int pos) { move(pos / getPositionsAmount(), pos % getPositionsAmount()); }
  int _getHole()
  {
    return getPositionsAmount() * getHoleBase() + getHolePosition();

    /* END SOLUTION */
  }

  /* END TEMPLATE */
}